File: 01-Programming-Languages/R/Data-Structures/LinkedList/linked_list_cycle.R
# Problem: Linked List Cycle Detection
# Description: Detect if a linked list has a cycle using Floyd's algorithm.
#
# Approach:
# - Use two pointers: slow (moves 1 step) and fast (moves 2 steps)
# - If they meet, a cycle exists
#
# Time Complexity: O(n)
# Space Complexity: O(1)
#
# Example:
# Input: 1 -> 2 -> 3 -> 4 -> 2 (cycle back to node 2)
# Output: TRUE

ListNode <- function(val = 0, next_node = NULL) {
  list(val = val, next = next_node)
}

has_cycle <- function(head) {
  if (is.null(head) || is.null(head$next)) return(FALSE)
  slow <- head
  fast <- head$next
  while (!is.null(slow) && !is.null(fast) && !is.null(fast$next)) {
    if (slow == fast) return(TRUE)
    slow <- slow$next
    fast <- fast$next$next
  }
  FALSE
}

# Create a linked list with a cycle: 1 -> 2 -> 3 -> 4 -> 2
n4 <- ListNode(4, NULL)
n3 <- ListNode(3, n4)
n2 <- ListNode(2, n3)
n1 <- ListNode(1, n2)
n4$next <- n2  # create cycle

cat("Has cycle:", has_cycle(n1), "\n")

# List without cycle
n5 <- ListNode(5, NULL)
n4b <- ListNode(4, n5)
n3b <- ListNode(3, n4b)
n2b <- ListNode(2, n3b)
n1b <- ListNode(1, n2b)

cat("Has cycle (no cycle):", has_cycle(n1b), "\n")

File: 01-Programming-Languages/Scala/Data-Structures/LinkedList/LinkedListCycle.scala
// Problem: Linked List Cycle Detection
// Description: Detect if a linked list has a cycle using Floyd's algorithm.
//
// Approach:
// - Use two pointers: slow (moves 1 step) and fast (moves 2 steps)
// - If they meet, a cycle exists
//
// Time Complexity: O(n)
// Space Complexity: O(1)
//
// Example:
// Input: 1 -> 2 -> 3 -> 4 -> 2 (cycle back to node 2)
// Output: true

object LinkedListCycle {
  class ListNode(var val: Int = 0, var next: ListNode = null)

  def hasCycle(head: ListNode): Boolean = {
    if (head == null || head.next == null) return false
    var slow = head
    var fast = head.next
    while (fast != null && fast.next != null) {
      if (slow == fast) return true
      slow = slow.next
      fast = fast.next.next
    }
    false
  }

  def main(args: Array[String]): Unit = {
    // List with cycle: 1 -> 2 -> 3 -> 4 -> 2
    val n4 = new ListNode(4)
    val n3 = new ListNode(3, n4)
    val n2 = new ListNode(2, n3)
    val n1 = new ListNode(1, n2)
    n4.next = n2
    println(s"Has cycle: ${hasCycle(n1)}")

    // List without cycle
    val n5b = new ListNode(5)
    val n4b = new ListNode(4, n5b)
    val n3b = new ListNode(3, n4b)
    val n2b = new ListNode(2, n3b)
    val n1b = new ListNode(1, n2b)
    println(s"Has cycle (no cycle): ${hasCycle(n1b)}")
  }
}
